# M3T5_Daniel_Lucas_Delgado

## Ejercicio 1. Transformaciones básicas con map y lambda

Creación:
    - Parte de una lista de números enteros, por ejemplo: [1, 2, 3, 4, 5].
    - Utiliza map y lambda para generar:
        - Una lista con los cuadrados de los números.
        - Una lista con los dobles.
        - Una lista de cadenas con el formato "N=<valor>".

Experimentación:
    - Muestra las tres listas resultantes.
    - Comprueba que las longitudes de las tres listas coinciden con la de entrada.
    - (Opcional) Aplica una segunda transformación con map a la lista de cadenas para añadir un sufijo, por ejemplo "(ok)".

In [3]:
listaEnteros = [1, 2, 3, 4, 5]
# Lambda Function para comprobar longitudes entre listas
comprobacion = lambda lista1, lista2: f"Longitud entre listas es igual: {len(lista1) == len(lista2)}\n"

# Map y Lambda para Cuadrados
listaCuadrados = list(map(lambda num: num * num, listaEnteros))
print("listaCuadrados:", listaCuadrados)
print(comprobacion(listaEnteros, listaCuadrados))

# Map y Lambda para Dobles
listaDobles = list(map(lambda num: num*2, listaEnteros))
print("listaDobles:", listaDobles)
print(comprobacion(listaEnteros, listaDobles))

# Map y Lambda para formato "N=<valor>"
listaCadenas = list(map(lambda num: f'N = {num}', listaEnteros))
print(listaCadenas)
print(comprobacion(listaEnteros, listaCadenas))

# Map opcional de doble trasnformacion de la listaCadenas con sufijo ok
listaCadenas2 = list(map(lambda string: f"{string}(ok)", listaCadenas))
print(listaCadenas2)

listaCuadrados: [1, 4, 9, 16, 25]
Longitud entre listas es igual: True

listaDobles: [2, 4, 6, 8, 10]
Longitud entre listas es igual: True

['N = 1', 'N = 2', 'N = 3', 'N = 4', 'N = 5']
Longitud entre listas es igual: True

['N = 1(ok)', 'N = 2(ok)', 'N = 3(ok)', 'N = 4(ok)', 'N = 5(ok)']


### Estructura
Para cada uno de los ejemplos se ha realizado un bloque de código con cada creación y experimentación.
    -Se crea la listaEnteros.
    -Creamos comprobacion, que es una lambda que va a comprobar longitudes entre listas.
    -Creamos lo que nos piden (unión de lambda y map).
    -Imprimimos el resultado y la comprobacion de longitud

### Map y Lambda
En este caso queremos mostrar una lista como resultado al crear cada una de las transformaciones que nos piden. Para ello:
    -Usamos list(), lo que nos permite convertir a lista lo que arroje la lambda function.
    -Usamos map(), para iterar la lambda function tantas veces como mida la lista.
    -Finalmente establecemos la lambda function.

```python
listaQuePidan = list(map(lambda num: num * num, listaEnteros))
````

### Estructura
Para cada uno de los ejemplos se ha realizado un bloque de código con cada creación y experimentación:

- Se crea la `listaEnteros`.
- Creamos `comprobacion`, que es una lambda que va a comprobar longitudes entre listas.
- Creamos lo que nos piden (unión de lambda y map).
- Imprimimos el resultado y la comprobación de longitud.

### Map y Lambda
En este caso queremos mostrar una lista como resultado al crear cada una de las transformaciones que nos piden. Para ello:

- Usamos `list()`, lo que nos permite convertir a lista lo que arroje la lambda function.
- Usamos `map()`, para iterar la lambda function tantas veces como mida la lista.
- Finalmente establecemos la lambda function.

```python
listaQuePidan = list(map(lambda num: num * num, listaEnteros))
```


## Ejercicio 2. Filtrado con filter y selección con list comprehension
Creación:
- Crea una lista de alumnos representados como diccionarios con las claves: nombre y nota (ejemplo: 4.5, 6.0, 8.75).
- Usando filter y lambda, obtén solo los aprobados (nota ≥ 5).
- Con una list comprehension, extrae la lista de nombres de los aprobados.
- Con otra list comprehension, crea una lista con f"{nombre} - {nota_redondeada}" (nota redondeada a 1 decimal).

Experimentación:
- Muestra la lista original, la de aprobados y la de nombres.
- Verifica que ningún alumno con nota menor de 5 aparezca entre los aprobados.
- (Opcional) Indica cómo ordenarías la lista de aprobados por nota descendente.

In [ ]:
# CREACIÓN
alumnos = [{"nombre": "Ana", "nota": 4.5},
           {"nombre": "Luis", "nota": 6.0},
           {"nombre": "Rodrigo", "nota": 8.75}]
# Lista de solo aprobados
listaAprobados = list(filter(lambda alumno: alumno["nota"]>= 5, alumnos))
#List Comprehension con nombres de aprobados
listaNombres = [alumno["nombre"] for alumno in listaAprobados]
#List Comprehension con Formato f-String
listaFormateada = [f"{alumno['nombre']} - {round(alumno['nota'], 1)}" for alumno in alumnos if alumno['nota'] >= 5]
listaFormateada2 = [f"{alumno['nombre']} - {round(alumno['nota'], 1)}" for alumno in listaAprobados]

# Experimentación
print("Lista Original:", alumnos)
print("Lista Aprobados:", listaAprobados)
print("Lista Nombres:", listaNombres)
print("Lista Formateada: ", listaFormateada)
print("Lista Formateada: ", listaFormateada2)




Lista Original: [{'nombre': 'Ana', 'nota': 4.5}, {'nombre': 'Luis', 'nota': 6.0}, {'nombre': 'Rodrigo', 'nota': 8.75}]
Lista Aprobados: [{'nombre': 'Luis', 'nota': 6.0}, {'nombre': 'Rodrigo', 'nota': 8.75}]
Lista Nombres: ['Luis', 'Rodrigo']
Lista Formateada:  ['Luis - 6.0', 'Rodrigo - 8.8']
Lista Formateada:  ['Luis - 6.0', 'Rodrigo - 8.8']


### Creación:
- Se crea la lista de alumnos con diccionarios para cada alumo.
___

#### Lista Aprobados.
Primero convertimos el return de la lambda en lista envolviendo todo con list(). Despues usamos filter() para recoger los resultados aprobados. Por ultimo la lambda alumno vamos a especificar que queremos aquellos alumnos con una nota superior a 5.
```python
# Lista de solo aprobados
listaAprobados = list(filter(lambda alumno: alumno["nota"]>= 5, alumnos))
```
Este sería el resultado. Cabe recordar que estamos acceediendo a la lista alumnos pero dentro de estas hay diccionarios, por eso para acceder a la nota lo hacemos así 
```python
alumno["nota"]
```
___
#### Lista Nombres.
Para la lista de Aprobados podemos usar la listaAprobados, la cual contiene el nombre de los alumnos cuya nota es superior o igual a 5. Para ello usamos list comprehension:
- Para recorrer y extraer solo los nombres de los diccionarios lo hacemos igual que en el ejemplo anterior pero con la clave de "nombre".
- El orden es parte que queremos extraer, bucle for que extrae esos alumnos in listaAprobados.
```python
listaNombres = [alumno["nombre"] for alumno in listaAprobados]
```
___
#### Lista formateada.
Por último esta lista formateada usa list comprehension para añadir el nombre y la nota a un f-string.
- Como el f-string va a ser el resultado final va a ser el primer elemento.
- El bucle for alumno in alumnos, para extraer el nombre y la nota de los diccionarios en la lista Alumnos.
- La condición para entrar en este f-String es mostrar los aprobados es que alumno["nota"] sea superior o igual a 5.
```python
listaFormateada = [f"{alumno['nombre']} - {round(alumno['nota'], 1)}" for alumno in alumnos if alumno['nota'] >= 5]
```
Sin embargo, si nos damos cuenta podemos extraer esta información y ahorrarnos el if, siempre y cuando los datos mostrados de los aprobados este bien escrito. De esta manera transformamos la lista con list comprehension:
```python
listaFormateada2 = [f"{alumno['nombre']} - {round(alumno['nota'], 1)}" for alumno in listaAprobados]
```
___

## Ejercicio 3.Comprehension mixtas. Listas, conjuntos y diccionarios.
Creación:
- Parte de una lista de emails (ejemplo: "ana@empresa.com", "luis@uni.es", "ana@empresa.com", "marta@empresa.com").
- Con una list comprehension, extrae la lista de dominios (lo que va después de @).
- Con una set comprehension, obtén el conjunto de dominios únicos.
- Con una dict comprehension, crea un diccionario {dominio: cantidad_apariciones} contando cuántas veces aparece cada dominio.
Experimentación:
- Muestra la lista de dominios, el conjunto (sin duplicados) y el diccionario de frecuencias.
- (Opcional) Crea un diccionario que asigne a cada email su longitud.
- (Opcional) Crea una lista que filtre solo los emails con dominio "empresa.com".

In [2]:
# CREACIÓN
emails = ["ana@empresa.com", "luis@uni.es", "villa@empresa.com", "marta@empresa.com"]

# LCompr lista Dominios @...
listaDominios = [email.split("@")[1] for email in emails]

# Set Comprehension de dominios
setDominios = {dominio for dominio in listaDominios}

# Dict Comprehension cuantas veces aparece el dominio en la lista
dictDominios = {dominio: listaDominios.count(dominio) for dominio in setDominios}
# EXPERIMENTACION
print(listaDominios)
print(setDominios)
print(dictDominios)

# OPCIONAL
# Crea un diccionario que asigne a cada email su longitud.
lenEmail = {email: len(email) for email in emails}
print(lenEmail) # -> {'ana@empresa.com': 15, 'luis@uni.es': 11, 'marta@empresa.com': 17}

#Crea una lista que filtre solo los emails con dominio "empresa.com".
listaFiltrada = [email for email in emails if email.split("@")[1] == "empresa.com"]
print(listaFiltrada)

['empresa.com', 'uni.es', 'empresa.com', 'empresa.com']
{'empresa.com', 'uni.es'}
{'empresa.com': 3, 'uni.es': 1}
{'ana@empresa.com': 15, 'luis@uni.es': 11, 'villa@empresa.com': 17, 'marta@empresa.com': 17}
['ana@empresa.com', 'villa@empresa.com', 'marta@empresa.com']


### List comprehension: emails
Se nos esta pidiendo una lista de los dominios, esto es, lo que viene después de los emails. Podemos acceder a ello con:
```python
email = "dldevtech@gmail.com"
email.split("@")[1] # -> "gmail.com"
```
Esto dividira el string en dos partes desde la cadena @. Como solo queremos el dominio de cada email, la segunda parte de la lista con la frase dividida es la posición 1[1]. 
Después vendría el elemento, que sería la variable donde vamos a guardar lo que rescatemos de cada iteración del bucle:
```python
listaDominios = [email.split("@")[1] for email in iterador]
```
Con esto nos queda la parte iteradora que en nuestro caso es la lista de emails que nos han dado:
```python
listaDominios = [email.split("@")[1] for email in emails]
```
___
### Set comprehension: Conjunto de dominios únicos.
Todo lo que hemos aplicado para transformar listas podemos aplicarlo a los conjuntos. Lo único que varia es que para declararlo envolvemos todo ente llaves { }. Como ya tenemos una lista de los dominios de los emails, podemos usarlo como objetivo a transformar:
```python
setDominios = {dominio for dominio in listaDominios} # -> {'uni.es', 'empresa.com'}
```
En este caso, la expresión es dominio, que es lo que se añade al conjunto en cada iteración. Después especificamos que cada elemento sobre el que se itera es listaDominios y cada parte resultante, o el return de la expresión será dominio.
___
### Dict Comprehension: Apariciones dominio.
Otro tipo de colección que podemos transformar es diccionarios. Exactamente de la misma manera que lo hacemos con las listas o los conjuntos. Pero la estructura que seguimos en este caso es la siguiente:
```python
{clave: valor  for elemento in colección}
```
En un dict comprehension:

- La clave es lo que va antes de los dos puntos, en nuestro ejemplo dominio.
- El valor es lo que va después la cantidad de veces que aparece un dominio -> listaDominios.count(dominio).
- El elemento del bucle es lo que aparece después del for, dominio también porque en este caso es cada una de las partes del iterador.
- La colección es lo que se recorre, el iterador que decíamos, queremos hacerlo solo tantas veces como dominios haya, por eso necesitamos el conjunto, para evitar iteraciones duplicadas.
Resultado final:
```python
dictDominios = {dominio: listaDominios.count(dominio) for dominio in setDominios} # -> {'uni.es': 1, 'empresa.com': 3}
```
___
#### OPCIONAL1. Crea un diccionario que asigne a cada email su longitud.
Para calcular cuántos caractéres tiene un string, podemos hacerlo gracias a la función len(). Como en el caso anterior (Apariciones de dominio), la estructura será la misma:
- Clave: email, es decir va a ser lo que vamos a asignar como clave en nuestro diccionario, por ejemplo: "ana@empresa.com".
- Valor: el número de carácteres que tiene ese email en concreto. Para ello hacemos len(email).
- Elemento: Cada uno de los elementos del iterador, en este caso será cada email en concreto de cada usuario.
- Colección: El iterador que en nuestro caso es la lista original de emails sin transformar.

```python
lenEmail = {email: len(email) for email in emails}
```
___
#### OPCIONAL2. Crea una lista que filtre solo los emails con dominio "empresa.com".
Aquí he optado por obtener una listra filtrada gracias a la condicion de que sí despues de cortar el dominio la parte resultante es empresa.com.
- Primera variable email va a ser lo que se añada en cada iteración del bucle.
- Segunda variable despues del for, email, es el elemento que se va a obtener en el bucle al desarrollar el condicional.
- Tercera variable, in emails, porque obtenemos la lista original de emails.
- La condición if email.split("@")[1] == "empresa.com", lo que comentabamos al principio, queremos solo los resultados que al dividir el string coincida con el dominio que queramos, en este caso "empresa.com".
```python
#Crea una lista que filtre solo los emails con dominio "empresa.com".
listaFiltrada = [email for email in emails if email.split("@")[1] == "empresa.com"]
print(listaFiltrada)
```
___

## Ejercicio 4. Pipeline funcional con filter + map y alternativa con list comprehensions
Creación:
- Dispones de una lista de productos (diccionarios) con nombre, precio y stock.
- Usando filter + lambda, selecciona solo los productos con stock > 0 y precio ≥ 10.
- Usando map + lambda, aplica un 10% de descuento al precio de los filtrados y genera tuplas (nombre, precio_descuento).
- Con una list comprehension, crea una alternativa equivalente en una sola expresión que devuelva las mismas tuplas.
Experimentación:
- Muestra:
    - La lista original.
    - La lista filtrada.
    - El resultado tras map.
    - El resultado de la list comprehension equivalente.
- Verifica que ambas salidas coinciden.
- (Opcional) Calcula el total de la compra sumando los precios con descuento usando sum().
___

In [36]:
# TODO CREACIÓN
listaProductos = [{"nombre": "Queso Tierno", "precio": 2.40, "stock": 10},
                  {"nombre": "Percha x10", "precio": 0.99, "stock": 4},
                  {"nombre": "Cajonera", "precio": 35.50, "stock": 12},
                  {"nombre": "FunkoPopo", "precio": 15.45, "stock": 10},
                  {"nombre": "Melocotón", "precio": 1, "stock": 8}]

# filter + lambda, selecciona solo los productos con stock > 0 y precio ≥ 10.
filtroProductos = list(filter(lambda producto: producto["stock"] > 0 and producto["precio"] >= 10, listaProductos))
# map + lambda, aplica un 10% de descuento al precio de los filtrados y genera tuplas (nombre, precio_descuento).
descuentoProductos = list(map(lambda producto: (producto["nombre"], (producto["precio"] * 0.9)), filtroProductos))
# list comprehension, crea una alternativa equivalente en una sola expresión que devuelva las mismas tuplas.
descuentoComprehension =  [(producto['nombre'], producto["precio"] * 0.9) for producto in filtroProductos]
#TODO: (Opcional) Calcula el total de la compra sumando los precios con descuento usando sum().
lambdaF = sum([(producto['precio'] * 0.9) for producto in filtroProductos])
listComprehension =[(producto['precio'] * 0.9) for producto in filtroProductos]

# EXPERIMENTACION
print(listaProductos)
print(filtroProductos)
print(descuentoProductos)
print(descuentoComprehension)
print(lambdaF)
print(listComprehension)


[{'nombre': 'Queso Tierno', 'precio': 2.4, 'stock': 10}, {'nombre': 'Percha x10', 'precio': 0.99, 'stock': 4}, {'nombre': 'Cajonera', 'precio': 35.5, 'stock': 12}, {'nombre': 'FunkoPopo', 'precio': 15.45, 'stock': 10}, {'nombre': 'Melocotón', 'precio': 1, 'stock': 8}]
[{'nombre': 'Cajonera', 'precio': 35.5, 'stock': 12}, {'nombre': 'FunkoPopo', 'precio': 15.45, 'stock': 10}]
[('Cajonera', 31.95), ('FunkoPopo', 13.905)]
[('Cajonera', 31.95), ('FunkoPopo', 13.905)]
45.855
[31.95, 13.905]


### Filter + Lambda. FiltroProductos
Lo que nos estan pidiendo por lo general suele tener esta composición:
```python
filtro = list(filter(lambda elemento: primeraCondicionElemento and segundaCondicionElemento, listaObjetivo))
```
Siguiendo con la fórmula:
- elemento: producto. Un diccionario como: *{'nombre': 'Queso Tierno', 'precio': 2.4, 'stock': 10}*
- primeraCondicionElemento: *producto["stock"] > 0*
- segundaCondicionElemento: *producto["precio"] >= 10*
- listaObjetivo: *listaProductos*. La lista sobre la que vamos a filtrar.

```python
filtroProductos = list(filter(lambda producto: producto["stock"] > 0 and producto["precio"] >= 10, listaProductos))
```
___
### Map + Lambda. DescuentoProductos en Tuplas
- Estructura: *list(map(lambda x: x, iterador))*
- PrecioDescuento: (precio*0.90)
- Tupla. La lambda tiene  que exportar lo resultante en una tupla: ((producto["nombre"], producto["precio"*0.90]), iterador)
- Lista Objetivo o iterador: filtroProductos (apartado anterior):
```python
descuentoProductos = list(map(lambda producto: (producto["nombre"], producto["precio"] * 0.9), filtroProductos))
```
___
### list Comprehension para el Descuento a la lista filtroProductos en Tuplas.

- Estructura: *[(resultado) for elemento in iterador]*
- Resultado: (producto["nombre"], producto["precio"] * 0.9)
- Código:
```python
descuentoComprehension = [(producto["nombre"], producto["precio"] * 0.9) for producto in filtroProductos]
```
___
### (Opcional) Total de la compra sumando los precios con descuento usando sum().
- Estructura: *sum(iterador)*
- Iterador: list comprehension:
```python
listComprehension = [(producto['precio'] * 0.9) for producto in filtroProductos]
```
De esta manera podemos sumar los datos que obtengamos de una list coprehension anterior.
- Código:
```python
lambdaF = sum([(producto['precio'] * 0.9) for producto in filtroProductos])
```


## Ejercicio 5. Gestión de encuestas: limpieza y análisis funcional
Contexto: una empresa ha recogido respuestas de una encuesta de satisfacción en varias sedes. Cada sede devuelve una lista de puntuaciones (del 1 al 5). Tenemos una lista de listas con todas las respuestas.

### Ejemplo de datos:
- Sede A: [5, 4, 3, 5]
- Sede B: [2, 3, 3, 4]
- Sede C: [1, 5, 5, 2]

### Creación:
- Usa una list comprehension anidada para aplanar todos los resultados en una sola lista de puntuaciones.
- Usa una set comprehension para identificar las puntuaciones únicas recogidas (valores de 1 a 5).
- Usando filter + lambda, selecciona solo las respuestas altas (≥ 4).
- Usando map + lambda, transforma esas respuestas en cadenas del tipo "Excelente (5)" o "Buena (4)".
- (Opcional) Cambia el criterio de filtrado para analizar respuestas bajas (≤ 2) sin modificar la estructura funcional.
### Experimentación:
- Muestra la lista original (matriz de respuestas), la lista aplanada, el set de puntuaciones únicas, los filtrados altos y su transformación textual.
- Repite el proceso con el filtro de respuestas bajas y observa los resultados.